libyul/optimiser/SyntacticalEquality.cpp

```diff
@@ -84,25 +88,30 @@ bool SyntacticallyEqual::statementEqual(ExpressionStatement const& _lhs, Express
 }
 bool SyntacticallyEqual::statementEqual(Assignment const& _lhs, Assignment const& _rhs)
 {
-	return
-		compareVector<Identifier, &SyntacticallyEqual::expressionEqual>(_lhs.variableNames, _rhs.variableNames) &&
-		(*this)(*_lhs.value, *_rhs.value);
+	return containerEqual(_lhs.variableNames, _rhs.variableNames, [this](Identifier const& _lhs, Identifier const& _rhs) -> bool {
```

Comment:
There's no way to just use `containerEqual(_lhs.variableNames, _rhs.variableNames, SyntacticallyEqual::expressionEqual)`, is there?

Reply:
No, not if ``containerEqual`` is extracted. That's why I wrote ``compareVector`` as a member before, but that's rather inflexible as well (then I can't use lambdas e.g. to just dereference for the switch cases, but can **only** use member function pointers, unless I write several overloads which kind of defeats the purpose...).

I can't write ``containerEqual(_lhs.variableNames, _rhs.variableNames, &SyntacticallyEqual::expressionEqual)`` (the ampersand is needed in any case), since I'm missing the ``this`` pointer for calling ``expressionEqual``.

One could use ``std::bind(&SyntacticallyEqual::expressionEqual, this, std::placeholders::_1, std::placeholders::_2)``, but that's not particularly nice as well IMO and the placeholders are non-optional. There is ``std::bind_front(&SyntacticallyEqual::expressionEqual, this)``, but only in C++20... there used to be ``std::bind1st(&SyntacticallyEqual::expressionEqual, this)``, but that again is deprecated in C++11 and removed in C++17 :-).

So the standard is a bit messy here :-).

I'd be open to writing an appropriate and nice binding mechanism for member functions in libdevcore, though - I guess it might in fact be useful elsewhere...

Reply:
With this:
```
template<std::size_t... N, typename R, typename C, typename... Args, typename... BoundArgs>
auto bindFront(std::index_sequence<N...>, R(C::*_c)(Args...), C* _obj, BoundArgs&&... _args) {
    return [_obj, _c, _args = std::forward_as_tuple(_args...)](auto&&... tail) -> auto {
        return (_obj->*_c)(std::forward<std::tuple_element_t<N, decltype(_args)>>(std::get<N>(_args))..., std::forward<decltype(tail)>(tail)...);
    };
}

template<typename R, typename C, typename... Args, typename... BoundArgs, typename Indices = std::make_index_sequence<sizeof...(BoundArgs)>>
auto bindFront(R(C::*_c)(Args...), C* _obj, BoundArgs&&... _args) {
    return bindFront(Indices{}, _c, _obj, std::forward<BoundArgs>(_args)...);
}

```
we could write ``bindFront(&SyntacticallyEqual::expressionEqual, this)`` with zero overhead ;-). (I couldn't resist trying to write that, but I'm fine with leaving it as is)